pkg/kv/kvclient/rangefeed/rangefeed.go

```diff
@@ -427,54 +433,98 @@ func (f *RangeFeed) processEvents(
-				if f.onDeleteRange == nil {
-					if f.knobs != nil && f.knobs.IgnoreOnDeleteRangeError {
-						continue
-					}
-					return errors.AssertionFailedf(
-						"received unexpected rangefeed DeleteRange event with no OnDeleteRange handler: %s", ev)
-				}
-				f.onDeleteRange(ctx, ev.DeleteRange)
-			case ev.Metadata != nil:
-				if f.onMetadata == nil {
-					return errors.AssertionFailedf("received unexpected metadata event with no OnMetadata handler")
-				}
-				f.onMetadata(ctx, ev.Metadata)
-			case ev.Error != nil:
-				// Intentionally do nothing, we'll get an error returned from the
-				// call to RangeFeed.
+			if err := f.processEvent(ctx, frontier, ev.RangeFeedEvent, ev.RegisteredSpan); err != nil {
+				return err
 			}
 		case <-ctx.Done():
 			return ctx.Err()
 		}
 	}
 }
+
+func (f *RangeFeed) processEvent(
+	ctx context.Context, frontier span.Frontier, ev *kvpb.RangeFeedEvent, registeredSpan roachpb.Span,
+) error {
+	switch {
+	case ev.Val != nil:
+		f.onValue(ctx, ev.Val)
+	case ev.Checkpoint != nil:
+		ts := ev.Checkpoint.ResolvedTS
+		if f.frontierQuantize != 0 {
+			ts.Logical = 0
+			ts.WallTime -= ts.WallTime % int64(f.frontierQuantize)
+		}
+		advanced, err := frontier.Forward(ev.Checkpoint.Span, ts)
+		if err != nil {
+			return err
+		}
+		if f.onCheckpoint != nil {
+			f.onCheckpoint(ctx, ev.Checkpoint)
+		}
+		if advanced && f.onFrontierAdvance != nil {
+			f.onFrontierAdvance(ctx, frontier.Frontier())
+		}
+		if f.frontierVisitor != nil {
+			f.frontierVisitor(ctx, advanced, frontier)
+		}
+	case ev.SST != nil:
+		if f.onSSTable == nil {
+			return errors.AssertionFailedf(
+				"received unexpected rangefeed SST event with no OnSSTable handler")
+		}
+		f.onSSTable(ctx, ev.SST, registeredSpan)
+	case ev.DeleteRange != nil:
+		if f.onDeleteRange == nil {
+			if f.knobs != nil && f.knobs.IgnoreOnDeleteRangeError {
+				return nil
+			}
+			return errors.AssertionFailedf(
+				"received unexpected rangefeed DeleteRange event with no OnDeleteRange handler: %s", ev)
+		}
+		f.onDeleteRange(ctx, ev.DeleteRange)
+	case ev.Metadata != nil:
+		if f.onMetadata == nil {
+			return errors.AssertionFailedf("received unexpected metadata event with no OnMetadata handler")
+		}
+		f.onMetadata(ctx, ev.Metadata)
+	case ev.Error != nil:
+		// Intentionally do nothing, we'll get an error returned from the
+		// call to RangeFeed.
+	case ev.BulkEvents != nil:
+		if f.onValues != nil {
+			// We can optimistically assume the bulk event consists of all value
+			// events, and allocate a buffer for them to be passed to onValues. In the
+			// rare case we hit a non-value event (it would have to be a rannge key as
+			// only a catch-up scan curently produces bulk events), we will throw our
+			// this buffer and any events we might have copied to it so far and just
+			// fallback to to processing each event, but this should be so uncommon it
+			// is not worth worrying about the potential wasted work.
```

Comment:
```suggestion
			// We can optimistically assume the bulk event consists of all value
			// events, and allocate a buffer for them to be passed to onValues. In the
			// rare case we hit a non-value event (it would have to be a range key as
			// only a catch-up scan currently produces bulk events), we will throw out
			// this buffer and any events we might have copied to it so far and just
			// fallback to to processing each event, but this should be so uncommon it
			// is not worth worrying about the potential wasted work.
```